Add Screenshot Tests and a Screenshotbot Integration

tdrhq opened the pull request:
## :scroll: Description

We talked over Email, and I wanted to build out the skeleton of Screenshot tests. It's not ideal to use screenshot-tests-for-android with Espresso tests because it's hard to control for non-determinism, but in this case, since you already had many Espresso tests, I figured it was best to plug it in here. I built a nice ViewAction to make it easier to write these tests, perhaps I should upstream that to screenshot-tests-for-android.

## :green_heart: How did you test it?
So this was the tricky part. I tested it on my CircleCI account, but had to make many temporary modifications to the config.yml (gcloud bucket, google project etc.) So I didn't test on the exact version that's in this Pull Request, but I tested on a very close version. If any thing fails it should be easy to resolve the issues.

## :camera_flash: Screenshots / GIFs

Here's an example image generated by the new runs:
<img src="https://screenshotbot.io/image/blob/603e1fc6c4519f6338c0f153/default.png" />

To be clear this was generated from the tests. 

Extra notes:

Feel free to squash the commits. I had lots of tiny commits that I chose not to squash for now because it's easier for me to undo changes for testing on CircleCI.
Also, as mentioned in the email, you would have to add SCREENSHOTBOT_API_KEY and SCREENSHOTBOT_API_SECRET environment variables for the build to be green. If you also want to start seeing Screenshotbot notices on PRs (it won't show on this PR because it's the absolute first PR, so there's no baseline image to compare against) you will also need to install the Screenshotbot app from Github marketplace.

Finally, I'm always around for support or questions. I mentioned we're still small, but that means we can promise you priority and even bespoke support if you need it.
<!-- <img src="https://user-images.githubusercontent.com/883468/93824506-8ef15880-fc31-11ea-9f44-ca32bb0de88a.jpg" width="260"/> -->

tdrhq committed:
First version with some (non-deterministic) screenshots

tdrhq committed:
temporary switch of project name

tdrhq committed:
Hmm, Gcloud is complaining on CircleCI

tdrhq committed:
Hmm, Gcloud is complaining on CircleCI

tdrhq committed:
Hmm, Gcloud is complaining on CircleCI

tdrhq committed:
Tweaks

tdrhq committed:
bring in screenshotbot

tdrhq committed:
CircleCI test

tdrhq committed:
Change directory

tdrhq committed:
Checkout repo for screenshotbot

tdrhq committed:
...

tdrhq committed:
...

tdrhq committed:
...

tdrhq committed:
...

tdrhq committed:
...

tdrhq committed:
...

tdrhq committed:
one level less

tdrhq committed:
Undo some config.yml changes

tdrhq committed:
remove unwanted pull

tdrhq commented:
I just realized I had a broken image on the PR summary. Here's a better link with all the screenshots in the last run I did with Screenshotbot: https://screenshotbot.io/runs/603e2d09c4519f6338c0f159

alisonthemonster commented:
Thank you so much for taking the time to do this!! Especially for getting CI working just right, that takes patience! 

I added my API key for Screenshotbot and voila it worked! The only thing thats still failing is that it can't find the coverage reports because the device changed. You'll need to change one line then it should pass all three checks :)

alisonthemonster reviewed (commented):
This line needs updated here:
`command: mkdir -p app/build/outputs/code-coverage/connected && cp firebase/walleye-28-en_US-portrait/artifacts/coverage.ec app/build/outputs/code-coverage/connected/coverage.ec
` to be 27

tdrhq committed:
Update coverage API level.

tdrhq commented:
Updated, I should've mentioned in my PR: I switched the API level to 27 because of this https://github.com/facebook/screenshot-tests-for-android/issues/260. Hopefully, Firebase Test Lab would fix this limitation soon.

tdrhq commented on app/src/androidTest/java/journal/gratitude/com/gratitudejournal/ui/EntryFragmentInstrumentedTest.kt:
```diff
@@ -335,4 +358,20 @@ class EntryFragmentInstrumentedTest {
-}
\ No newline at end of file
+    // Make the inspiration message deterministic for screenshot tests
+    fun removeInspiration() {
+        // There's probably a less verbose way to do this correctly,
+        // but this works for now.
+        onView(withId(R.id.inspiration))
+            .perform(object: ViewAction {
+                         override fun getConstraints(): Matcher<View> = any(View::class.java)
+
+                         override fun getDescription() = "Make inspiration message deterministic"
+                         override fun perform(uiController: UiController, view: View) {
+                             val tv = view as TextView
+                             tv.setText("\"This is a wonderful day. I\'ve never seen this day before\" \nMaya Angelou");
+                         }
+            });
+    }
+
```
Btw, yesterday I added a new feature to deal with non-determinism like this in an easier manner. Instead of always trying to make screenshots deterministic in the Screenshotbot UI you can select "Edit Masks" and highlight all the areas that aren't going to be deterministic (so for instance, in this case, you'd draw a rectangle around where the quote would be). Screenshotbot would then disregard any changes in those areas.

alisonthemonster replied:
Very cool! I will look into the masking :)

tdrhq commented:
If I'm reading this correctly, the latest test_instrumented failure seems to be from my CircleCI builds (the ones I set up for testing this PR). I've disabled my job and I'll make a dummy change to reattempt the build on your account.

tdrhq committed:
Specify that the main branch is
